### PR TITLE
simplify grad clip tests

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #8642 

Refactor four tests into two tests:

1)  combine these two tests to one:
```
test_gradient_clipping
test_gradient_clipping_fp16
```

combine these two tests to one:
```
test_gradient_clipping_by_value
test_gradient_clipping_by_value_fp16
```

2) perform the asserts in `FitLoop.backward` instead of `FitLoop.training_step_and_backward`, which is equally valid (required for #8642 )

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: tests/trainer/test_trainer.py [hunk 1]
```diff
@@ -873,148 +873,74 @@ def on_keyboard_interrupt(self, trainer, pl_module):
...
-    trainer.fit(model)
-
-
-def test_gradient_clipping_by_value(tmpdir):
-    """
-    Test gradient clipping by value
-    """
+@pytest.mark.parametrize(
+    "precision,gpus",
+    [pytest.param(32, 0), pytest.param(16, 1, marks=RunIf(min_gpus=1, amp_native=True))],
```
Comment (Borda) [inline tests/trainer/test_trainer.py:878]:
```suggestion
    [(32, 0), pytest.param(16, 1, marks=RunIf(min_gpus=1, amp_native=True))],
```

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -873,148 +873,74 @@ def on_keyboard_interrupt(self, trainer, pl_module):
...
 
 
-@RunIf(min_gpus=1, amp_native=True)
-def test_gradient_clipping_by_value_fp16(tmpdir):
-    """
-    Test gradient clipping by value with fp16
-    """
+@pytest.mark.parametrize(
+    "precision,gpus",
+    [pytest.param(32, 0), pytest.param(16, 1, marks=RunIf(min_gpus=1, amp_native=True))],
```
Comment (Borda) [inline tests/trainer/test_trainer.py:911]:
```suggestion
    [(32, 0), pytest.param(16, 1, marks=RunIf(min_gpus=1, amp_native=True))],
```

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -873,148 +873,74 @@ def on_keyboard_interrupt(self, trainer, pl_module):
...
-    grad_clip_val = 1e-10
+    model = EvalModelTemplate()  # TODO: when precision=16, BoringModel produces NaN, but EvalModelTemplate not
     trainer = Trainer(
+        default_root_dir=tmpdir,
         max_steps=1,
         max_epochs=1,
-        gradient_clip_val=grad_clip_val,
-        gradient_clip_algorithm="value",
-        default_root_dir=tmpdir,
+        gpus=gpus,
```
Comment (Borda) [inline tests/trainer/test_trainer.py:889]:
```suggestion
        gpus=int(torch.cuda.is_available()),
```
then we do not need the one argument :]

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -873,148 +873,74 @@ def on_keyboard_interrupt(self, trainer, pl_module):
...
     model = BoringModel()
+
     grad_clip_val = 1e-10
     trainer = Trainer(
         max_steps=1,
         max_epochs=1,
-        precision=16,
-        gpus=1,
+        precision=precision,
+        gpus=gpus,
```
Comment (Borda) [inline tests/trainer/test_trainer.py:924]:
```suggestion
        gpus=int(torch.cuda.is_available())
```

**Review by Borda — APPROVED** [review]

File: tests/trainer/test_trainer.py (hunk 1 above)
Comment (carmocca) [inline tests/trainer/test_trainer.py:878]:
```suggestion
    [(32, 0), pytest.param(16, 1, marks=RunIf(min_gpus=1, amp_native=True))],
```

File: tests/trainer/test_trainer.py (hunk 2 above)
Comment (carmocca) [inline tests/trainer/test_trainer.py:911]:
```suggestion
    [(32, 0), pytest.param(16, 1, marks=RunIf(min_gpus=1, amp_native=True))],
```

**Review by carmocca — APPROVED** [review]

File: tests/trainer/test_trainer.py (hunk 4 above)
Comment (awaelchli) [reply tests/trainer/test_trainer.py]: done, and removed gpus argument
thanks

**Review by kaushikb11 — APPROVED** [review]
